src/resources/rmt_esp32.cc

```diff
@@ -17,459 +17,611 @@
-  return Smi::from(threshold);
+MODULE_IMPLEMENTATION(rmt, MODULE_RMT);
+
+PRIMITIVE(bytes_per_memory_block) {
+  return Smi::from(SOC_RMT_MEM_WORDS_PER_CHANNEL * sizeof(word));
 }
 
-PRIMITIVE(set_idle_threshold) {
-  ARGS(RmtResource, resource, uint16, threshold)
-  esp_err_t err = rmt_set_rx_idle_thresh(resource->channel(), threshold);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  return process->null_object();
+PRIMITIVE(init) {
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) FAIL(ALLOCATION_FAILED);
+
+  RmtResourceGroup* rmt = _new RmtResourceGroup(process, EventQueueEventSource::instance());
+  if (!rmt) FAIL(MALLOC_FAILED);
+
+  proxy->set_external_address(rmt);
+  return proxy;
 }
 
-PRIMITIVE(config_bidirectional_pin) {
-  ARGS(int, pin, RmtResource, resource, bool, enable_pullup);
+PRIMITIVE(channel_new) {
+  ARGS(RmtResourceGroup, resource_group, int, pin_num, uint32, resolution, uint32, block_symbols, int, kind)
 
-  // This function has to be called after the pin was first
-  // configured as an output channel, and then as an input channel.
-  // We now need to reenable the output without losing the input.
+  if (block_symbols == 0) FAIL(INVALID_ARGUMENT);
 
-  // Enable the pin. "w1ts" = "write 1 to set".
-  // TODO(florian): not completely sure why this is needed, but without it
-  // it won't work.
-#if defined(CONFIG_IDF_TARGET_ESP32C3) || defined(CONFIG_IDF_TARGET_ESP32C6)
-  if (pin >= MAX_GPIO_NUM) FAIL(INVALID_ARGUMENT);
-  GPIO.enable_w1ts.enable_w1ts = (0x1 << pin);
-#else
-  if (pin < 32) {
-    GPIO.enable_w1ts = (0x1 << pin);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) FAIL(ALLOCATION_FAILED);
+
+  bool handed_to_resource = false;
+
+  const int caps_flags = RMT_MEM_ALLOC_CAPS;
+  auto resource_memory = heap_caps_malloc(sizeof(RmtResource), caps_flags);
+  if (!resource_memory) FAIL(MALLOC_FAILED);
+  Defer free_resource_memory { [&] { if (!handed_to_resource) free(resource_memory); } };
+
+  bool is_tx = kind != 0;
+
+  RmtInOut* in_out;
+  if (is_tx) {
+    auto out_memory = heap_caps_malloc(sizeof(RmtOut), caps_flags);
+    if (!out_memory) FAIL(MALLOC_FAILED);
+    in_out = new (out_memory) RmtOut();
   } else {
-    GPIO.enable1_w1ts.data = (0x1 << (pin - 32));
+    auto in_memory = heap_caps_malloc(sizeof(RmtIn), caps_flags);
+    if (!in_memory) FAIL(MALLOC_FAILED);
+    in_out = new (in_memory) RmtIn();
   }
-#endif
-  esp_err_t err = rmt_set_gpio(resource->channel(), RMT_MODE_TX, static_cast<gpio_num_t>(pin), false);
+  Defer free_in_out { [&] { if (!handed_to_resource) delete in_out; } };
+
+  // No need for a big queue. We only allow one read/write at a time.
+  QueueHandle_t queue = xQueueCreate(1, sizeof(word));
+  if (queue == null) FAIL(MALLOC_FAILED);
+  Defer free_queue { [&] { if (!handed_to_resource) vQueueDelete(queue); } };
+
+  rmt_channel_handle_t handle;
+  esp_err_t err;
+  if (kind == 0) {
```

Comment:
Maybe just use `is_tx` here? If you always want to deal with rx first, then maybe have `is_rx` instead and use a positive test?

Reply:
changed the order, so that tx is always done first.